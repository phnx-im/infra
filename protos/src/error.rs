use std::fmt;

use tonic::Status;

#[derive(Debug, thiserror::Error)]
#[error("missing field: {field}")]
pub struct MissingFieldError<E: fmt::Display> {
    field: E,
}

impl<E: fmt::Display> From<MissingFieldError<E>> for Status {
    fn from(e: MissingFieldError<E>) -> Self {
        Status::invalid_argument(e.to_string())
    }
}

pub trait MissingFieldExt<T, E: fmt::Display> {
    fn ok_or_missing_field(self, field: E) -> Result<T, MissingFieldError<E>>;
}

impl<T, E: fmt::Display> MissingFieldExt<T, E> for Option<T> {
    fn ok_or_missing_field(self, field: E) -> Result<T, MissingFieldError<E>> {
        self.ok_or(MissingFieldError { field })
    }
}

pub struct InvalidTls<F> {
    error: tls_codec::Error,
    field: F,
}

impl<F: fmt::Display> From<InvalidTls<F>> for Status {
    fn from(InvalidTls { error, field }: InvalidTls<F>) -> Status {
        Status::invalid_argument(format!("Invalid TLS field {field}: {error}"))
    }
}

pub struct TlsFailed<F> {
    error: tls_codec::Error,
    field: F,
}

impl<F: fmt::Display> From<TlsFailed<F>> for Status {
    fn from(TlsFailed { error, field }: TlsFailed<F>) -> Status {
        Status::internal(format!("TLS serialization of {field} failed: {error}",))
    }
}

pub trait InvalidTlsExt {
    type Value;

    fn invalid_tls<F>(self, field: F) -> Result<Self::Value, InvalidTls<F>>;

    fn tls_failed<F>(self, field: F) -> Result<Self::Value, TlsFailed<F>>;
}

impl<T> InvalidTlsExt for Result<T, tls_codec::Error> {
    type Value = T;

    fn invalid_tls<F>(self, field: F) -> Result<T, InvalidTls<F>> {
        self.map_err(|error| InvalidTls { error, field })
    }

    fn tls_failed<F>(self, field: F) -> Result<T, TlsFailed<F>> {
        self.map_err(|error| TlsFailed { error, field })
    }
}
