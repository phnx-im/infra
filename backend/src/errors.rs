// SPDX-FileCopyrightText: 2024 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use thiserror::Error;
use tracing::error;

#[derive(Debug, Error)]
pub enum StorageError {
    #[error(transparent)]
    Database(#[from] DatabaseError),
    #[error("Error deserializing column: {0}")]
    Serde(#[from] phnxtypes::codec::Error),
}

impl From<sqlx::Error> for StorageError {
    fn from(e: sqlx::Error) -> Self {
        Self::Database(e.into())
    }
}

impl From<Box<dyn std::error::Error + Send + Sync>> for StorageError {
    fn from(e: Box<dyn std::error::Error + Send + Sync>) -> Self {
        Self::Database(e.into())
    }
}

impl From<StorageError> for tonic::Status {
    fn from(e: StorageError) -> Self {
        match e {
            StorageError::Database(error) => {
                error!(%error, "Database error");
                Self::internal("Database error")
            }
            StorageError::Serde(error) => {
                error!(%error, "Seriazation error");
                Self::internal("Seriazation error")
            }
        }
    }
}

#[derive(Debug, Error)]
pub enum DatabaseError {
    #[error(transparent)]
    Sqlx(#[from] sqlx::Error),
    #[error(transparent)]
    Dynamic(#[from] Box<dyn std::error::Error + Send + Sync>),
}

/// General error while accessing the requested queue.
#[derive(Error, Debug)]
pub(super) enum QueueError {
    #[error(transparent)]
    Storage(#[from] StorageError),
    /// Mismatching sequence numbers.
    #[error("Mismatching sequence numbers.")]
    SequenceNumberMismatch,
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
}

impl From<sqlx::Error> for QueueError {
    fn from(e: sqlx::Error) -> Self {
        Self::Storage(e.into())
    }
}

impl From<phnxtypes::codec::Error> for QueueError {
    fn from(e: phnxtypes::codec::Error) -> Self {
        Self::Storage(e.into())
    }
}
