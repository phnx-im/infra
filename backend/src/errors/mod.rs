// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use displaydoc::Display;
use mls_assist::{
    group, memory_provider::MlsAssistMemoryStorage, openmls::group::MergeCommitError,
};
use thiserror::Error;
use tokio::sync::mpsc;
use tonic::Status;
use tracing::error;

use aircommon::codec::PersistenceCodec;

pub(crate) mod auth_service;
pub(crate) mod qs;

pub(crate) type CborMlsAssistStorage = MlsAssistMemoryStorage<PersistenceCodec>;

#[derive(Debug, Error)]
pub enum StorageError {
    #[error(transparent)]
    Database(#[from] DatabaseError),
    #[error("Error deserializing column: {0}")]
    Serde(#[from] aircommon::codec::Error),
}

impl From<sqlx::Error> for StorageError {
    fn from(e: sqlx::Error) -> Self {
        Self::Database(e.into())
    }
}

impl From<Box<dyn std::error::Error + Send + Sync>> for StorageError {
    fn from(e: Box<dyn std::error::Error + Send + Sync>) -> Self {
        Self::Database(e.into())
    }
}

impl From<StorageError> for Status {
    fn from(error: StorageError) -> Self {
        error!(%error, "storage error");
        match error {
            StorageError::Database(_) => Self::internal("Database error"),
            StorageError::Serde(_) => Self::internal("Seriazation error"),
        }
    }
}
#[derive(Debug, Error)]
pub enum DatabaseError {
    #[error(transparent)]
    Sqlx(#[from] sqlx::Error),
    #[error(transparent)]
    Dynamic(#[from] Box<dyn std::error::Error + Send + Sync>),
}

/// General error while accessing the requested queue.
#[derive(Error, Debug, Display)]
pub(super) enum QueueError {
    /// {0:?}
    Storage(#[from] StorageError),
    /// Mismatching sequence numbers
    SequenceNumberMismatch,
    /// Payload receiver closed
    PayloadReceiverClosed,
}

impl From<sqlx::Error> for QueueError {
    fn from(e: sqlx::Error) -> Self {
        Self::Storage(e.into())
    }
}

impl From<aircommon::codec::Error> for QueueError {
    fn from(e: aircommon::codec::Error) -> Self {
        Self::Storage(e.into())
    }
}

impl<T> From<mpsc::error::SendError<T>> for QueueError {
    fn from(_: mpsc::error::SendError<T>) -> Self {
        Self::PayloadReceiverClosed
    }
}

impl From<QueueError> for Status {
    fn from(error: QueueError) -> Self {
        let msg = error.to_string();
        match error {
            QueueError::Storage(error) => {
                error!(%error, "storage error");
                Self::internal(msg)
            }
            QueueError::SequenceNumberMismatch => Self::internal(msg),
            QueueError::PayloadReceiverClosed => Self::internal(msg),
        }
    }
}

/// Potential errors when performing a group operation.
#[derive(Debug, Error)]
pub(crate) enum GroupOperationError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message")]
    ProcessingError,
    /// Missing queue config in client key package.
    #[error("Missing queue config in client key package")]
    MissingQueueConfig,
    /// Incomplete Welcome message.
    #[error("Incomplete Welcome message.")]
    IncompleteWelcome,
    #[error("Error merging commit")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<GroupOperationError> for Status {
    fn from(e: GroupOperationError) -> Self {
        let msg = e.to_string();
        match e {
            GroupOperationError::LibraryError
            | GroupOperationError::ProcessingError
            | GroupOperationError::InvalidMessage => Status::internal(msg),
            GroupOperationError::MissingQueueConfig | GroupOperationError::IncompleteWelcome => {
                Status::invalid_argument(msg)
            }
            GroupOperationError::MergeCommitError(merge_commit_error) => {
                error!(%merge_commit_error, "failed merging commit");
                Status::internal(msg)
            }
        }
    }
}

/// Potential errors when joining a connection group.
#[derive(Debug, Error)]
pub(crate) enum JoinConnectionGroupError {
    /// Invalid assisted message.
    #[error("Invalid assisted message")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message")]
    ProcessingError,
    /// Not a connection group.
    #[error("Not a connection group")]
    NotAConnectionGroup,
    #[error("Error merging commit")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<JoinConnectionGroupError> for Status {
    fn from(e: JoinConnectionGroupError) -> Self {
        let msg = e.to_string();
        match e {
            JoinConnectionGroupError::InvalidMessage
            | JoinConnectionGroupError::NotAConnectionGroup => Status::invalid_argument(msg),
            JoinConnectionGroupError::ProcessingError => Status::internal(msg),
            JoinConnectionGroupError::MergeCommitError(merge_commit_error) => {
                error!(%merge_commit_error, "failed merging commit");
                Status::internal(msg)
            }
        }
    }
}

/// Potential errors when deleting a group.
#[derive(Debug, Error)]
pub(crate) enum GroupDeletionError {
    /// Invalid assisted message.
    #[error("Invalid assisted message")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message")]
    ProcessingError,
    #[error("Error merging commit")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<GroupDeletionError> for Status {
    fn from(e: GroupDeletionError) -> Self {
        let msg = e.to_string();
        match e {
            GroupDeletionError::InvalidMessage => Status::invalid_argument(msg),
            GroupDeletionError::ProcessingError => Status::internal(msg),
            GroupDeletionError::MergeCommitError(merge_commit_error) => {
                error!(%merge_commit_error, "failed merging commit");
                Status::internal(msg)
            }
        }
    }
}

/// Potential errors when processing a self remove proposal.
#[derive(Debug, Error)]
pub(crate) enum ClientSelfRemovalError {
    /// Invalid assisted message.
    #[error("Invalid assisted message")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message")]
    ProcessingError,
    #[error("Error merging commit")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<ClientSelfRemovalError> for Status {
    fn from(e: ClientSelfRemovalError) -> Self {
        let msg = e.to_string();
        match e {
            ClientSelfRemovalError::InvalidMessage => Status::invalid_argument(msg),
            ClientSelfRemovalError::ProcessingError => Status::internal(msg),
            ClientSelfRemovalError::MergeCommitError(merge_commit_error) => {
                error!(%merge_commit_error, "failed merging commit");
                Status::internal(msg)
            }
        }
    }
}

/// Potential errors when resyncing a client.
#[derive(Debug, Error)]
pub(crate) enum ResyncClientError {
    /// Invalid assisted message.
    #[error("Invalid assisted message")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message")]
    ProcessingError,
    #[error("Error merging commit")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<ResyncClientError> for Status {
    fn from(e: ResyncClientError) -> Self {
        let msg = e.to_string();
        match e {
            ResyncClientError::InvalidMessage => Status::invalid_argument(msg),
            ResyncClientError::ProcessingError => Status::internal(msg),
            ResyncClientError::MergeCommitError(merge_commit_error) => {
                error!(%merge_commit_error, "failed merging commit");
                Status::internal(msg)
            }
        }
    }
}
