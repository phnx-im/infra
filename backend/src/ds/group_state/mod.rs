// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use std::collections::BTreeMap;

use mls_assist::{
    MlsAssistRustCrypto,
    group::Group,
    openmls::{
        group::GroupId,
        prelude::{GroupEpoch, LeafNodeIndex},
        treesync::RatchetTree,
    },
    provider_traits::MlsAssistProvider,
};
use phnxtypes::{
    codec::PhnxCodec,
    crypto::{
        ear::{
            Ciphertext, EarDecryptable, EarEncryptable,
            keys::{EncryptedIdentityLinkKey, GroupStateEarKey},
        },
        errors::{DecryptionError, EncryptionError},
    },
    errors::CborMlsAssistStorage,
    identifiers::{QsReference, SealedClientReference},
    messages::client_ds::WelcomeInfoParams,
    time::TimeStamp,
};
use serde::{Deserialize, Serialize};
use sqlx::PgExecutor;
use thiserror::Error;
use tracing::error;
use uuid::Uuid;

use crate::errors::StorageError;

use super::{GROUP_STATE_EXPIRATION, ReservedGroupId, process::ExternalCommitInfo};

pub(super) mod persistence;

#[derive(Debug, Serialize, Deserialize)]
pub(super) struct MemberProfile {
    pub(super) leaf_index: LeafNodeIndex,
    pub(super) encrypted_identity_link_key: EncryptedIdentityLinkKey,
    pub(super) client_queue_config: QsReference,
    pub(super) activity_time: TimeStamp,
    pub(super) activity_epoch: GroupEpoch,
}

/// The `DsGroupState` is the per-group state that the DS persists.
/// It is encrypted-at-rest with a roster key.
///
/// TODO: Past group states are now included in mls-assist. However, we might
/// have to store client credentials externally.
pub struct DsGroupState {
    pub(super) group: Group,
    pub(super) provider: MlsAssistRustCrypto<PhnxCodec>,
    pub(super) member_profiles: BTreeMap<LeafNodeIndex, MemberProfile>,
}

impl DsGroupState {
    pub(crate) fn new(
        provider: MlsAssistRustCrypto<PhnxCodec>,
        group: Group,
        creator_encrypted_identity_link_key: EncryptedIdentityLinkKey,
        creator_queue_config: QsReference,
    ) -> Self {
        let creator_client_profile = MemberProfile {
            encrypted_identity_link_key: creator_encrypted_identity_link_key,
            client_queue_config: creator_queue_config,
            activity_time: TimeStamp::now(),
            activity_epoch: 0u64.into(),
            leaf_index: LeafNodeIndex::new(0u32),
        };
        let client_profiles = [(LeafNodeIndex::new(0u32), creator_client_profile)].into();
        Self {
            provider,
            group,
            member_profiles: client_profiles,
        }
    }

    /// Get a reference to the public group state.
    pub(crate) fn group(&self) -> &Group {
        &self.group
    }

    /// Get a mutable reference to the public group state.
    pub(crate) fn group_mut(&mut self) -> &mut Group {
        &mut self.group
    }

    pub(super) fn welcome_info(
        &mut self,
        welcome_info_params: WelcomeInfoParams,
    ) -> Option<&RatchetTree> {
        self.group_mut()
            .past_group_state(&welcome_info_params.epoch, &welcome_info_params.sender)
    }

    pub(super) fn external_commit_info(&self) -> ExternalCommitInfo {
        let group_info = self.group().group_info().clone();
        let ratchet_tree = self.group().export_ratchet_tree();
        let encrypted_identity_link_keys = self.encrypted_identity_link_keys();
        ExternalCommitInfo {
            group_info,
            ratchet_tree,
            encrypted_identity_link_keys,
        }
    }

    /// Create a vector of encrypted identity link keys from the current list of
    /// client records.
    pub(super) fn encrypted_identity_link_keys(&self) -> Vec<EncryptedIdentityLinkKey> {
        self.member_profiles
            .values()
            .map(|client_profile| client_profile.encrypted_identity_link_key.clone())
            .collect()
    }

    pub(super) fn encrypt(
        self,
        ear_key: &GroupStateEarKey,
    ) -> Result<EncryptedDsGroupState, DsGroupStateEncryptionError> {
        let encrypted =
            EncryptableDsGroupState::from(SerializableDsGroupState::from_group_state(self)?)
                .encrypt(ear_key)?;
        Ok(encrypted)
    }

    pub(super) fn decrypt(
        encrypted_group_state: &EncryptedDsGroupState,
        ear_key: &GroupStateEarKey,
    ) -> Result<Self, DsGroupStateDecryptionError> {
        let encryptable = EncryptableDsGroupState::decrypt(ear_key, encrypted_group_state)?;
        let group_state = SerializableDsGroupState::into_group_state(encryptable.into())?;
        Ok(group_state)
    }
}

#[derive(Debug, Error)]
pub(super) enum DsGroupStateEncryptionError {
    #[error("Error decrypting group state: {0}")]
    EncryptionError(#[from] EncryptionError),
    #[error("Error deserializing group state: {0}")]
    DeserializationError(#[from] phnxtypes::codec::Error),
}

impl From<DsGroupStateEncryptionError> for tonic::Status {
    fn from(value: DsGroupStateEncryptionError) -> Self {
        match value {
            DsGroupStateEncryptionError::EncryptionError(error) => {
                error!(%error, "Failed to encrypt group state");
                Self::internal("Failed to encrypt group state")
            }
            DsGroupStateEncryptionError::DeserializationError(error) => {
                error!(%error, "Failed to deserialize group state");
                Self::internal("Failed to deserialize group state")
            }
        }
    }
}

#[derive(Debug, Error)]
pub(super) enum DsGroupStateDecryptionError {
    #[error("Error decrypting group state: {0}")]
    DecryptionError(#[from] DecryptionError),
    #[error("Error deserializing group state: {0}")]
    DeserializationError(#[from] phnxtypes::codec::Error),
}

impl From<DsGroupStateDecryptionError> for tonic::Status {
    fn from(error: DsGroupStateDecryptionError) -> Self {
        match error {
            DsGroupStateDecryptionError::DecryptionError(error) => {
                error!(%error, "Failed to decrypt group state");
                Self::internal("Failed to decrypt group state")
            }
            DsGroupStateDecryptionError::DeserializationError(error) => {
                error!(%error, "Failed to deserialize group state");
                Self::internal("Failed to deserialize group state")
            }
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq, Eq)]
#[serde(transparent)]
pub struct EncryptedDsGroupState(Ciphertext);

#[derive(Debug)]
#[cfg_attr(test, derive(PartialEq, Eq))]
pub(super) struct StorableDsGroupData {
    group_id: Uuid,
    pub(super) encrypted_group_state: EncryptedDsGroupState,
    last_used: TimeStamp,
    deleted_queues: Vec<SealedClientReference>,
}

impl StorableDsGroupData {
    pub(super) async fn new_and_store<'a>(
        connection: impl PgExecutor<'a>,
        group_id: ReservedGroupId,
        encrypted_group_state: EncryptedDsGroupState,
    ) -> Result<Self, StorageError> {
        let group_data = Self {
            group_id: group_id.0,
            encrypted_group_state,
            last_used: TimeStamp::now(),
            deleted_queues: vec![],
        };
        group_data.store(connection).await?;
        Ok(group_data)
    }

    pub(super) fn has_expired(&self) -> bool {
        self.last_used.has_expired(GROUP_STATE_EXPIRATION)
    }
}

impl From<Ciphertext> for EncryptedDsGroupState {
    fn from(ciphertext: Ciphertext) -> Self {
        Self(ciphertext)
    }
}

impl AsRef<Ciphertext> for EncryptedDsGroupState {
    fn as_ref(&self) -> &Ciphertext {
        &self.0
    }
}

#[derive(Serialize, Deserialize)]
pub(crate) struct SerializableDsGroupState {
    group_id: GroupId,
    serialized_provider: Vec<u8>,
    member_profiles: Vec<(LeafNodeIndex, MemberProfile)>,
}

impl SerializableDsGroupState {
    pub(super) fn from_group_state(
        group_state: DsGroupState,
    ) -> Result<Self, phnxtypes::codec::Error> {
        let group_id = group_state
            .group()
            .group_info()
            .group_context()
            .group_id()
            .clone();
        let client_profiles = group_state.member_profiles.into_iter().collect();
        let serialized_provider = group_state.provider.storage().serialize()?;
        Ok(Self {
            group_id,
            serialized_provider,
            member_profiles: client_profiles,
        })
    }

    pub(super) fn into_group_state(self) -> Result<DsGroupState, phnxtypes::codec::Error> {
        let storage = CborMlsAssistStorage::deserialize(&self.serialized_provider)?;
        // We unwrap here, because the constructor ensures that `self` always stores a group
        let group = Group::load(&storage, &self.group_id)?.unwrap();
        let client_profiles = self.member_profiles.into_iter().collect();
        let provider = MlsAssistRustCrypto::from(storage);
        Ok(DsGroupState {
            provider,
            group,
            member_profiles: client_profiles,
        })
    }
}

#[derive(Serialize, Deserialize)]
pub(super) enum EncryptableDsGroupState {
    V1(SerializableDsGroupState),
}

impl From<EncryptableDsGroupState> for SerializableDsGroupState {
    fn from(encryptable: EncryptableDsGroupState) -> Self {
        match encryptable {
            EncryptableDsGroupState::V1(serializable) => serializable,
        }
    }
}

impl From<SerializableDsGroupState> for EncryptableDsGroupState {
    fn from(serializable: SerializableDsGroupState) -> Self {
        EncryptableDsGroupState::V1(serializable)
    }
}

impl EarEncryptable<GroupStateEarKey, EncryptedDsGroupState> for EncryptableDsGroupState {}
impl EarDecryptable<GroupStateEarKey, EncryptedDsGroupState> for EncryptableDsGroupState {}
