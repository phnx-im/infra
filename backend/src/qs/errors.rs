// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use crate::errors::{QueueError, StorageError};

use super::network_provider::NetworkProvider;
use aircommon::crypto::errors::{DecryptionError, EncryptionError, KeyGenerationError};
use thiserror::Error;
use tracing::error;

// === DS API errors ===

/// Error fetching a message from the QS.
#[derive(Error, Debug)]
pub enum QsEnqueueError<N: NetworkProvider> {
    /// Unseal error
    #[error(transparent)]
    UnsealError(#[from] DecryptionError),
    /// An error ocurred enqueueing in a fan out queue
    #[error(transparent)]
    EnqueueError(#[from] EnqueueError),
    /// An error ocurred while sending a message to the network
    #[error("An error ocurred while sending a message to the network")]
    NetworkError(N::NetworkError),
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid response
    #[error("Invalid response")]
    InvalidResponse,
}

/// Error enqueuing a fanned-out message.
#[derive(Error, Debug)]
pub enum EnqueueError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError, // E.g. an error while encoding a message before enqueing it.
    /// Error in the underlying storage provider
    #[error("Error in the underlying storage provider")]
    Storage(StorageError),
    /// Client not found
    #[error("Client not found")]
    ClientNotFound,
    /// Queue error
    #[error("Queue error")]
    Queue(QueueError),
}

impl From<StorageError> for EnqueueError {
    fn from(error: StorageError) -> Self {
        error!(%error, "Failed to enqueue message due to storage error");
        EnqueueError::Storage(error)
    }
}

impl From<sqlx::Error> for EnqueueError {
    fn from(error: sqlx::Error) -> Self {
        error!(%error, "Failed to enqueue message due to sqlx error");
        EnqueueError::Storage(error.into())
    }
}

impl From<QueueError> for EnqueueError {
    fn from(error: QueueError) -> Self {
        error!(%error, "Failed to enqueue message due to queue error");
        match error {
            QueueError::Storage(storage_error) => storage_error.into(),
            QueueError::SequenceNumberMismatch | QueueError::PayloadReceiverClosed => {
                Self::Queue(error)
            }
        }
    }
}

impl From<EncryptionError> for EnqueueError {
    fn from(error: EncryptionError) -> Self {
        error!(%error, "Failed to enqueue message due to encryption error");
        EnqueueError::LibraryError
    }
}

// === Internal errors ===

#[derive(Debug, Error)]
pub(super) enum GenerateAndStoreError {
    #[error("Error generating signature keypair")]
    KeyGenerationError(#[from] KeyGenerationError),
    #[error("Error storing key")]
    StorageError(#[from] StorageError),
}
