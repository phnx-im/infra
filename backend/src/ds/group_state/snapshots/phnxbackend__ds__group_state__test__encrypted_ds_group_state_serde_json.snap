---
source: backend/src/ds/group_state/mod.rs
expression: state
---
{
  "ciphertext": [
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1
  ],
  "nonce": [
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1
  ]
}
